SETUP/upgrade/09/01_daily_page_limits.php

```diff
@@ -0,0 +1,39 @@
+<?php
+
+$relPath='../../../pinc/';
+include($relPath.'connect.inc');
+new dbConnect();
+
+header('Content-type: text/plain');
+
+
+echo "Creating 'daily_page_limits' table...\n";
+$sql = "
+    CREATE TABLE daily_page_limit
+    (
+        name             VARCHAR(50) NOT NULL,
+        project_selector VARCHAR(256) NOT NULL,
+        daily_limit      INT(8) NOT NULL,
+        PRIMARY KEY (name)
+    )
+";
+echo "$sql\n";
+mysqli_query(DPDatabase::get_connection(), $sql) or die( mysqli_error(DPDatabase::get_connection()) );
+
+
+echo "Creating 'daily_page_limit_user_counts' table...\n";
```

Comment:
Why can't we use the TallyBoard for this instead of creating yet another table to track pretty much the same thing?

Reply:
Because it's only "pretty much" the same thing, not the same thing. In particular, the F1 limit has to ignore pages saved in the F1 self-evaluation project.

Reply:
But is it "close enough" to not matter? If we put this in people will then want to see how close they are to a given limit which means we have to put it in some page for them to view. It cascades quickly.

Reply:
Basically, can we go back to Linda et al and say "hey, the code is simpler if we implement a per-round per-person per-day limit". Because then we can use the TallyBoard and we don't need an `eval()` at all. Yes, it's less flexible but it's also less complex and will probably still address the core need.

Reply:
There was also a request to allow for the possibility of having a higher daily limit for specific projects (e.g. projects that have been available longer, or special day projects). And I think there was also talk of using this mechanism to limit pages in beginner projects. So none of the anticipated use cases are as simple as "the number of pages saved in a given round".

Reply:
Of course not, nothing we do is simple.

Then this needs a way for a user to see what limits impact them and how close they are to those limits. It also needs a crontab or something to reset the contents of this table, presumably at midnight.

Reply:
>  this needs a way for a user to see what limits impact them

See the new code in project.php. Whenever they visit a project that contributes to a daily page limit, the "blurbs" will tell them so, and also tell them their current count toward that limit.

> It also needs a crontab or something to reset the contents of this table, presumably at midnight.

Yup. (For testing, we just "simulated midnight" whenever.)

Reply:
> See the new code in project.php. Whenever they visit a project that contributes to a daily page limit, the "blurbs" will tell them so, and also tell them their current count toward that limit.

That will be a poor user experience. A user will stop being able to proof a project and get redirected back to a project page where they are told they can't work on it anymore. Then they hunt around to try and find another project where they can work on, clicking through projects until they find one they can work on. If there are multiple limits in place this could be a very frustrating process.

If would be far better if the limits, and how closer the user was to them, was available somewhere for the user to look at -- such as on their stats page.

Reply:
> Then they hunt around to try and find another project where they can work on, clicking through projects until they find one they can work on.

I suspect there wouldn't be *that* much hunting involved. If you hit the limit for "X" projects, then you say to yourself "Okay, no more X projects for me today." So if you click on a project, thinking you'll be able to work on it, only to be told that you've already hit your limit for "X" projects, then either:
* you simply forgot (in which case, now you've been reminded), or
* you didn't realize that this project is an "X" project (contributes to the "X" daily limit). This is plausible, depending on how well the name of the limit conveys the set of projects that it covers (E.g., for "F1", it's pretty clear, but (say) "F1 older" and "F1 newer" wouldn't tell you exactly where the cutoff is.)

So that suggests a couple things:
* A DailyPageLimit should have a precise human-readable description of the project-selector that can be shown to the user when appropriate (e.g., as a tooltip or at the end of a link).
* In project listings, it might be helpful to mark (somehow) those projects that contribute to a limit that the user has already maxed out today.

> If would be far better if the limits, and how closer the user was to them, was available somewhere for the user to look at -- such as on their stats page.

(Do you mean "if instead" or "if in addition"?) Yup, something like that would be good. It would help people who've forgotten if/which limits they've reached today. But apart from that, I don't think it would improve the experience of 'hunting' for a project they can work on.

Reply:
Yes - in addition to the project pages.

Reply:
> > If would be far better if the limits, and how closer the user was to them, was available somewhere for the user to look at -- such as on their stats page.
> 
> (Do you mean "if instead" or "if in addition"?) Yup, something like that would be good. It would help people who've forgotten if/which limits they've reached today. But apart from that, I don't think it would improve the experience of 'hunting' for a project they can work on.

What about the "My Projects" page?

Reply:
It sort-of relates, in that your counts-for-today would all be from projects on your "Recent" tab, but I don't think it really fits, because it's not a list of projects. Conceivably it *could* be a list of projects (i.e., the ones where your saves today contributed to your current count for limit X), but that wasn't requested. (It would require either different record-keeping or some slightly dubious queries.)

Reply:
One of the problems with the user's Stats (or Member Details) page, assuming that's what is meant by stats page, is that it's not quick and easy to get to. The user has to go to Statistics Central and type their username into the search box (or is there an easier way that I've just missed for the last 10 years? :D). I've no problem with using that page, but it might be a good idea to make it more easily and obviously accessible, possibly with a direct link to "My Stats" at the top of Statistics Central?

Reply:
I think the daily limits contents belongs on the user's stats page. Yes, I've wanted to update how users access their own stats for a while, but I think that should be done outside of this PR.

Reply:
That's reasonable -- I just thought that it was worth mentioning. :)

Reply:
One other place that might be good for your-progress-toward-daily-limits is in the stats-bar on round pages. E.g., the F1 limit would make sense on the F1 page, under "Personal Statistics", say. (So each DailyPageLimit would have to know the round(s) for which it's relevant.)

Reply:
> (or is there an easier way that I've just missed for the last 10 years? :D).

(If you're on a round page, you can go to the sidebar, under "Personal Statistics", and click the "<round-id> Details..." link. From there you can go to your other round-specific stats pages, but I don't think there's a link back to the round-agnostic stats page where your access chart is, though you can get there by stripping the round param from the URL.)

Reply:
> > (or is there an easier way that I've just missed for the last 10 years? :D).
> 
> (If you're on a round page, you can go to the sidebar, under "Personal Statistics", and click the "[round-id] Details..." link. From there you can go to your other round-specific stats pages, but I don't think there's a link back to the round-agnostic stats page where your access chart is, though you can get there by stripping the `tally_name` parameter from the URL.)

Thanks. I've probably tripped over that occasionally in the past, but hadn't remembered it.